node-it/src/test/scala/com/wavesplatform/it/asset/InvokeReissueBurnAssetSuite.scala

```diff
@@ -0,0 +1,321 @@
+       |
+       |@Callable (i) func reissueAsset(a: ByteVector, r: Boolean, q: Int) = [Reissue(a, r, q)]
+       |
+       |$function
+       |
+       """.stripMargin
+  }
+
+  val simpleNonreissuableAsset = Map(
+    "type"           -> "Simple",
+    "compiledScript" -> None,
+    "name"           -> "SimpleAsset",
+    "description"    -> "description",
+    "quantity"       -> 100500,
+    "isReissuable"   -> false,
+    "decimals"       -> 8,
+    "nounce"         -> 0
+  )
+
+  val simpleReissuableAsset = Map(
+    "type"           -> "Reissuable",
+    "compiledScript" -> None,
+    "name"           -> "ReissuableAsset",
+    "description"    -> "description",
+    "quantity"       -> 100000000,
+    "isReissuable"   -> true,
+    "decimals"       -> 3,
+    "nounce"         -> 0
+  )
+
+  val nftAsset = Map(
+    "type"           -> "NFT",
+    "compiledScript" -> None,
+    "name"           -> "NFTAsset",
+    "description"    -> "description",
+    "quantity"       -> 1,
+    "isReissuable"   -> false,
+    "decimals"       -> 0,
+    "nounce"         -> 0
+  )
+
+  for (method <- Seq("@Callable", "Transaction")) s"Asset Issue/Reissue/Burn via $method" - {
+
+    for (data <- Seq(simpleNonreissuableAsset, simpleReissuableAsset, nftAsset)) s"${data("type")} asset could be issued in callable" in {
+      val acc = createDapp(script(data))
+
+      val fee = invocationCost(if (isNft(data)) 0 else 1)
+      val tx  = issue(acc, method, data, fee)
+
+      validateIssuedAssets(acc, tx, data, method = method)
+      sender.balanceDetails(acc).regular shouldBe (initialWavesBalance - setScriptPrice - fee)
+    }
+
+    for (data <- Seq(simpleNonreissuableAsset, simpleReissuableAsset)) s"${data("type")} asset could be partially burned" in {
+      val acc            = createDapp(script(data))
+      val fee            = invocationCost(if (isNft(data)) 0 else 1)
+      val txIssue        = issue(acc, method, data, fee)
+      val assetId        = validateIssuedAssets(acc, txIssue, data, method = method)
+      val burnQuantity   = 1000
+      val remainQuantity = data("quantity").asInstanceOf[Number].longValue - burnQuantity
+
+      burn(acc, method, assetId, burnQuantity)
+
+      sender.assetsDetails(assetId).quantity shouldBe remainQuantity
+      sender.assertAssetBalance(acc, assetId, remainQuantity)
+    }
+
+    for (data <- Seq(simpleNonreissuableAsset, simpleReissuableAsset, nftAsset)) s"${data("type")} could be fully burned" in {
+      val acc     = createDapp(script(data))
+      val fee     = invocationCost(if (isNft(data)) 0 else 1)
+      val txIssue = invokeScript(acc, "issueAsset", fee = fee)
+      val assetId = validateIssuedAssets(acc, txIssue, data, method = method)
+
+      burn(acc, method, assetId, data("quantity").asInstanceOf[Number].intValue)
+
+      sender.assetsDetails(assetId).quantity shouldBe 0
+      sender.assertAssetBalance(acc, assetId, 0)
+    }
+
+    "Reissuable asset could be reissued" in {
```

Comment:
what if the resulting quantity is greater than Long.MAX_VALUE?

Reply:
What if try to reissue/burn a smart asset that restricts such actions?

Reply:
what if the resulting quantity is greater than Long.MAX_VALUE? -- error, I've added the test

Reply:
What if try to reissue/burn a smart asset that restricts such actions? -- it should produce an error